pgmpy/inference/mplp.py

```diff
@@ -0,0 +1,670 @@
+from math import log
+import itertools
+import numpy as np
+from operator import sub
+from operator import add
+import queue
+import math
+
+
+class MPLPAlg:
+    """
+    Class for performing inference using Max-Product Linear Programming Relaxations.
+    """
+
+    def __init__(self, all_cardinalities, all_factors, all_lambdas):
+
+        # For containing the cardinalities of all the variables present in the model
+        self.m_var_sizes = all_cardinalities
+        # All the factors are kept here
+        self.m_all_intersect = all_factors
+        # All the messages per factor is kept here. It is updated in each iteration of run_mplp
+        self.objective = []
+        # It contain the Region objects for all the factors which are made of more than one variable
+        self.m_all_regions = []
+        # Initial messages for more than single node regions are kept here. These don't change.
+        self.m_region_lambdas = []
+        # Maximum integral solution of the single nodes makes the m_best_val
+        self.m_best_val = -1000000
+        # This is to be used for calculating the breaking point of RunMPLP
+        self.last_obj = 1e40
+        self.m_intersect_map = []
+        for i in range(len(all_factors)):
+            self.objective.append([log(y) for y in all_lambdas[i]])
+            if len(all_factors[i]) > 1:
+                self.m_region_lambdas.append(self.objective[i])
+                curr_region = Region(all_factors[i], all_factors[:i+1], all_factors[i], all_cardinalities, i)
+                self.m_all_regions.append(curr_region)
+            if len(all_factors[i]) == 2:
+                self.m_intersect_map.append((all_factors[i], i))
+
+            # Initialise output vector
+            self.m_decoded_res = [0]*len(all_cardinalities)
+            self.m_best_decoded_res = [0]*len(all_cardinalities)
+
+    def _get_flat_ind(self, base_ind, base_card):
```

Comment:
This method should not be present in inference class. It should be a part of the ```Factor``` class.